src/main/java/edu/cornell/kfs/module/ar/document/web/struts/CuContractsGrantsInvoiceDocumentAction.java

```diff
@@ -41,6 +56,115 @@ public ActionForward prorateBill(ActionMapping mapping, ActionForm form, HttpSer
         return super.prorateBill(mapping, form, request, response);
     }
 
+    @Override
+    public ActionForward route(ActionMapping mapping, ActionForm form, HttpServletRequest request, HttpServletResponse response) throws Exception {
+        ContractsGrantsInvoiceDocument contractsGrantsInvoiceDocument = ((ContractsGrantsInvoiceDocumentForm) form).getContractsGrantsInvoiceDocument();
+
+        String warningMessage = getContractsGrantsInvoiceDocumentWarningMessage(contractsGrantsInvoiceDocument);
+        if (StringUtils.isNotEmpty(warningMessage)) {
+            ActionForward forward = promptForFinalBillConfirmation(mapping, form, request, response, KFSConstants.ROUTE_METHOD, warningMessage, contractsGrantsInvoiceDocument);
+            if (forward != null) {
+                return forward;
+            }
+        }
+
+        return super.route(mapping, form, request, response);
+    }
+
+    protected String getContractsGrantsInvoiceDocumentWarningMessage(ContractsGrantsInvoiceDocument contractsGrantsInvoiceDocument) {
```

Comment:
This method seems to have too many of the validation pieces located directly within it. You may want to create separate helper methods for some of the validation sections if possible.